backends-velox/src/main/scala/org/apache/spark/sql/execution/BroadcastUtils.scala

```diff
@@ -40,23 +42,46 @@ object BroadcastUtils {
       from: Broadcast[F],
       fn: Iterator[ColumnarBatch] => Iterator[InternalRow]): Broadcast[T] = {
     mode match {
-      case HashedRelationBroadcastMode(_, _) => // no-op
+      case HashedRelationBroadcastMode(_, _) =>
+        // ColumnarBuildSideRelation to HashedRelation.
+        val fromBroadcast = from.asInstanceOf[Broadcast[ColumnarBuildSideRelation]]
+        val fromRelation = fromBroadcast.value.asReadOnlyCopy()
+        var rowCount: Long = 0
+        val toRelation = TaskResources.runUnsafe {
+          val rowIterator = fn(fromRelation.deserialized.flatMap {
+            cb =>
+              rowCount += cb.numRows()
+              Iterator(cb)
+          })
+          mode.transform(rowIterator, Some(rowCount))
+        }
+        // Rebroadcast Spark relation.
+        context.broadcast(toRelation).asInstanceOf[Broadcast[T]]
+      case IdentityBroadcastMode =>
+        // ColumnarBuildSideRelation to HashedRelation.
+        val fromBroadcast = from.asInstanceOf[Broadcast[ColumnarBuildSideRelation]]
+        val fromRelation = fromBroadcast.value.asReadOnlyCopy()
+        val toRelation = TaskResources.runUnsafe {
+          val rowIterator = fn(fromRelation.deserialized)
+          val rowArray = new ArrayBuffer[InternalRow]
+
+          /**
+           * [[io.glutenproject.execution.VeloxColumnarToRowExec.toRowIterator()]] creates a single
+           * UnsafeRow. The iterator uses this same unsafe row and keep on changing the pointer to
+           * point to new value. If we directly call rowIterator.toArray() then all the elements in
+           * array points to same UnsafeRow object resulting in wrong output. here we need to create
+           * a array having individual UnsafeRow object.
+           */
+          while (rowIterator.hasNext) {
+            val unsafeRow = rowIterator.next().asInstanceOf[UnsafeRow]
+            rowArray.append(unsafeRow.copy())
```

Comment:
@zhztheplayer can you please review this change and suggest if there is any better way to resolve it?

Reply:
`Array[InternalRow]` requires all rows to be materialized so I think it's fine to use `.copy()` in the initial version of this feature. We can still find a way to optimize the code later.

Reply:
And thank you for adding the detailed explanation in comment :)